Sources/SwiftLexicalLookup/ScopeImplementations.swift

```diff
@@ -96,28 +96,80 @@ import SwiftSyntax
   /// - for `memberBlockUpToLastDecl` - a, b, c, d
   /// - for `memberBlock` - a, b, c, d, e, f
   /// - for `codeBlock` - a
-  @_spi(Experimental) public func lookup(
-    for name: String?,
+  @_spi(Experimental) public func _lookup(
+    for identifier: Identifier?,
     at syntax: SyntaxProtocol,
-    with config: LookupConfig
+    with config: LookupConfig,
+    state: LookupState
   ) -> [LookupResult] {
-    let names = introducedNames(using: config.fileScopeHandling)
-      .filter { introducedName in
-        introducedName.isAccessible(at: syntax) && (name == nil || introducedName.refersTo(name!))
+    switch config.fileScopeHandling {
+    case .memberBlock:
+      let names = introducedNames(using: .memberBlock)
+        .filter { lookupName in
+          checkName(identifier, refersTo: lookupName, at: syntax)
+        }
+
+      return names.isEmpty ? [] : [.fromFileScope(self, withNames: names)]
+    case .memberBlockUpToLastDecl:
+      var members = [LookupName]()
+      var sequentialItems = [CodeBlockItemSyntax]()
+      var encounteredNonDeclaration = false
+
+      for codeBlockItem in statements {
+        let item = codeBlockItem.item
+
+        if encounteredNonDeclaration {
+          sequentialItems.append(codeBlockItem)
+        } else {
+          if item.is(DeclSyntax.self) || item.is(VariableDeclSyntax.self) {
+            let foundNames = LookupName.getNames(from: item)
+
+            members.append(contentsOf: foundNames.filter { checkName(identifier, refersTo: $0, at: syntax) })
+          } else {
+            encounteredNonDeclaration = true
+            sequentialItems.append(codeBlockItem)
+          }
+        }
       }
 
-    return [.fromFileScope(self, withNames: names)]
+      let sequentialNames = sequentialLookup(
+        in: sequentialItems,
+        for: identifier,
+        at: syntax,
+        with: config,
+        state: state,
+        createResultsForThisScopeWith: { .fromFileScope(self, withNames: $0) }
+      )
+
+      return (members.isEmpty ? [] : [.fromFileScope(self, withNames: members)]) + sequentialNames
+    }
   }
 }
 
-@_spi(Experimental) extension CodeBlockSyntax: ScopeSyntax {
+@_spi(Experimental) extension CodeBlockSyntax: SequentialScopeSyntax {
   /// Names introduced in the code block scope
   /// accessible after their declaration.
   @_spi(Experimental) public var introducedNames: [LookupName] {
     statements.flatMap { codeBlockItem in
       LookupName.getNames(from: codeBlockItem.item, accessibleAfter: codeBlockItem.endPosition)
     }
   }
+
+  @_spi(Experimental) public func _lookup(
+    for identifier: Identifier?,
+    at syntax: SyntaxProtocol,
+    with config: LookupConfig,
+    state: LookupState
+  ) -> [LookupResult] {
```

Comment:
Open question:

I find it quite confusing that we always have `self`, which is a syntax node and `syntax`, which  is also a syntax node. As far as I can tell the purpose of `syntax` is to define the position at which lookup occurs so that declarations introduced afterwards are not included. If this is the case, would it be clearer if `at` was an `AbsolutePosition`? If not, do you think some other name than `syntax` could make the distinction from `self` clearer?

Reply:
My intent to provide syntax node instead of just position was to have some flexibility in case we want to use e.g. it's `endPosition` instead of `position`. But you're right we're not using it for anything more than just accessing `position` currently. What would you say about [renaming it to `origin`](https://github.com/swiftlang/swift-syntax/blob/252d3be80c010add2bac9f9712ac32d9892ea574/Sources/SwiftLexicalLookup/ScopeSyntax.swift#L66).

Reply:
> My intent to provide syntax node instead of just position was to have some flexibility in case we want to use e.g. it's `endPosition` instead of `position`.

Wouldn’t that be an argument for making the parameter an `AbsolutePosition`? If we want to use `endPosition` instead of `position` in some cases, we would just need to change it at the location that calls `lookup` instead of at every location that does a position conversion to `origin`.

Reply:
I was thinking more of the API deciding whether to use `position` or `endPosition`, but more I think about now, the less I actually like this idea since it would be unnecessarily complex and ambiguous for clients. I changed it now to be just `AbsolutePosition`.